### PR TITLE
Feature: Implement .transform(X) in the lasagne wrapper to extract hidden representations

### PR DESCRIPTION
**Pull request by bmilde**
This allows you to call .transform(X) on a trained model (lasagne wrapper) to retrieve a transformed feature representation at an intermediate layer of your choice. Particularly useful to inspect your learned features or to stack your model with a different classifier. The functional MNIST test has been extended with an example of the latter, with a SVM on top of the neural network, using its last hidden representation as transformation layer. (This increases accuracy here from 0.877 to 0.903.)

### CODE REVIEW

**Comment by dnouri** [comment]
I like the idea a lot, but was wondering if the implementation could maybe be done in a way that all necessary changes are local to the `transform` method (apart from the parameter to `__init__`).

Another point is that if we're going to store `transform_iter_` on self, we'll have to remove it in `__getstate__`.

**Comment by bmilde** [comment]
The current implementation closely follows how predict(X) is implemented with self.predict_iter_, so I think it makes sense to thread transform(X) equally and precompute and store the iterator. Especially since a major use case for using the neural net as a feature extractor is chaining it to another classifier, so that the same reasoning that applies to predict(X) (e.g. performance) should also apply to transform(X).

The proposed implementation has one downside, the transformation layer has to be chosen when the NeuralNet is created. But this similar to a RandomForestClassifier in Sklearn, which can transform features to keep the n best features and n is chosen at initialisation. We could have a update_transform_layer(name) if that is a problem.

Yep, I missed __getstate__, should be included there. And while we're at it, self._transform_layer should probably be renamed to self.transform_layer_.

**Comment by bmilde** [comment]
Apparently Travis encountered a time-out (?). What is the maximum execution length and was that ever a problem? Scaling down form 5 epochs to 4 epochs in the NN or limiting the number of samples could help to scale down on testing time. I could also chain something else than a SVM, but the NN+SVM combo is what you can find in literature, so its a good example.

**Comment by edz504** [comment]
Is this slated to be merged in soon?  If not, is there a different way to get the output of an intermediate layer in the interim?  Thanks!

**Comment by dnouri** [comment]
@edz504: Look for "Theano layer functions and Feature Extraction" [in this tutorial](http://blog.christianperone.com/2015/08/convolutional-neural-networks-and-feature-extraction-with-python/).

**Comment by edz504** [comment]
Thanks for the quick and informative response as always!

**Comment by dnouri** [comment]
Sorry for leaving this stale for such a long time.

The equivalent is now implemented by #196, though it's using the method name `get_output` and you have to pass in the name of the layer explicitly, rather than having that a parameter of `NeuralNet`.  The speed concerns were also addressed.
